pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -189,25 +190,76 @@ func injectApmTelemetryConfig(pod *corev1.Pod) {
+// - if language detection is on and can detect the apps' languages, returns only auto-detected languages
+// - otherwise returns all tracing libraries supported by APM Instrumentation
+func getLibrariesToInjectForApmInstrumentation(pod *corev1.Pod, registry string) ([]libInfo, bool) {
+	autoDetected := false
+
+	// Pinned tracing libraries in APM Instrumentation configuration
+	libsToInject := getPinnedLibraries(registry)
+	if len(libsToInject) > 0 {
+		return libsToInject, autoDetected
+	}
+
+	// Tracing libraries for language detection
+	libsToInject = getLibrariesLanguageDetection(pod, registry)
+	if len(libsToInject) > 0 {
+		autoDetected = true
+		return libsToInject, autoDetected
+	}
+
+	// Latest tracing libraries for all supported languages (java, js, dotnet, python, ruby)
+	libsToInject = getAllLatestLibraries(registry)
+
+	return libsToInject, autoDetected
+}
+
+// getPinnedLibraries returns tracing libraries to inject as configured by apm_config.instrumentation.lib_versions
+func getPinnedLibraries(registry string) []libInfo {
+	libsToInject := []libInfo{}
 	var libVersion string
-	singleStepLibraryVersions := config.Datadog.GetStringMapString("apm_config.instrumentation.lib_versions")
 
-	for _, lang := range supportedLanguages {
-		libVersion = "latest"
+	singleStepLibraryVersions := config.Datadog.GetStringMapString("apm_config.instrumentation.lib_versions")
 
-		if version, ok := singleStepLibraryVersions[string(lang)]; ok {
-			log.Warnf("Library version %s is specified for language %s", version, string(lang))
+	// If APM Instrumentation is enabled and configuration apm_config.instrumentation.lib_versions specified, inject only the libraries from the configuration
+	for lang, version := range singleStepLibraryVersions {
+		if !slices.Contains(supportedLanguages, language(lang)) {
+			log.Warnf("APM Instrumentation detected configuration for unsupported language: %s. Tracing library for %s will not be injected", lang, lang)
+		} else {
+			log.Infof("Library version %s is specified for language %s", version, lang)
 			libVersion = version
+			libsToInject = append(libsToInject, libInfo{lang: language(lang), image: libImageName(registry, language(lang), libVersion)})
 		}
+	}
+	return libsToInject
+}
 
-		libsToInject[string(lang)] = libInfo{
-			lang:  lang,
-			image: libImageName(registry, lang, libVersion),
+// getLibrariesLanguageDetection runs process language auto-detection and returns languages to inject for APM Instrumentation.
+// The langages information is available in workloadmeta-store and attached on the pod's owner.
+func getLibrariesLanguageDetection(pod *corev1.Pod, registry string) []libInfo {
+	if config.Datadog.GetBool("admission_controller.run_language_detection") {
+		// Always run language detection to collect telemetry data
+		libsToInject := getAutoDetectedLibraries(pod, registry)
+		if config.Datadog.GetBool("admission_controller.inject_auto_detected_libraries") {
```

Comment:
Can we check both the conditions in top `if` .

Reply:
We want to always run language detection if `admission_controller.run_language_detection=true`, but only make the detected libraries available for injection if `admission_controller.inject_auto_detected_libraries=true`. How will it work with both conditions checked together?

Reply:
Hey @liliyadd 

I don't believe `run_language_detection` should be an admission controller config option.
Language detection is not an admission controller feature, it is a global feature. 
The admission controller only uses the outcome of the language detection feature.

We already have a config option to enable/disable running detection. (Check it [here](https://github.com/DataDog/datadog-agent/blob/070f8d1fb2afbe904ee933283cefca107350663b/pkg/config/setup/config.go#L1272-L1274))

I think it is sufficient to reuse this config option instead of creating a new one.

WDYT?

Reply:
Agree, we should exist existing configuration. From last week's sync I got an impression that this configuration doesn't exist yet.